### PR TITLE
Eliminate unnecessary specialized Reaction types

### PR DESCRIPTION
**Pull request by speth**
With the introduction of the `ReactionRate` class, most of the behavior that is specialized for particular kinds of reactions is contained in those classes, but we still have a (more or less) 1:1 correspondence with classes derived from `Reaction`, even though many of them don't have any unique behavior.

**Changes proposed in this pull request**

- Eliminate unnecessary specializations of C++ `Reaction` class.
  - So far, this covers `ElementaryReaction`, `PlogReaction`, and `BlowersMaselReaction`, and should be easily expandable to cover `CustomFunc1Reaction` and `TwoTempPlasmaReaction`.
  - The C++ `ThreeBodyReaction`, `FalloffReaction` and `ChebyshevReaction` do implement some specialized behaviors, mostly around parsing and representing the reaction equation, so I think they need to stay for now.

- In the Python module, my goal is to eliminate all the specializations of `Reaction`, at least for homogeneous phase reactions (I think `InterfaceReaction` will probably need to stay)
  - This would be more or less analogous to how we handle the `ThermoPhase` class in Python -- with a few exceptions, there aren't wrappers for different `ThermoPhase` types. Right now, I'm a bit stuck based on one of the routes that's been introduced for constructing `Reaction` objects.

**If applicable, provide an example illustrating new features this pull request is introducing**

Before this PR, a `PlogReaction` could be created in one step as:
```python
R = PlogReaction(
    equation="H2 + O2 <=> 2 OH",
    rate=[(1013.25, Arrhenius(1.2124e+16, -0.5779, 45491376.8)),
          (101325., Arrhenius(4.9108e+31, -4.8507, 103649395.2)),
          (1013250., Arrhenius(1.2866e+47, -9.0246, 166508556.0)),
          (10132500., Arrhenius(5.9632e+56, -11.529, 220076726.4))],
    kinetics=gas)
```
Or in a two steps:
```python
R = PlogReaction({"H2": 1, "O2": 1}, {"OH": 2})
R.rate = PlogRate(
    [(1013.25, Arrhenius(1.2124e+16, -0.5779, 45491376.8)),
     (101325., Arrhenius(4.9108e+31, -4.8507, 103649395.2)),
     (1013250., Arrhenius(1.2866e+47, -9.0246, 166508556.0)),
     (10132500., Arrhenius(5.9632e+56, -11.529, 220076726.4))]
```

With this PR, the first example becomes:
```python
R = Reaction(
    equation="H2 + O2 <=> 2 OH",
    rate=PlogRate([(1013.25, Arrhenius(1.2124e+16, -0.5779, 45491376.8)),
          (101325., Arrhenius(4.9108e+31, -4.8507, 103649395.2)),
          (1013250., Arrhenius(1.2866e+47, -9.0246, 166508556.0)),
          (10132500., Arrhenius(5.9632e+56, -11.529, 220076726.4))]),
    kinetics=gas)
```
which is pretty similar.

The second case becomes:
```python
R = ct.Reaction({"H2": 1, "O2": 1}, {"OH": 2})
R.rate = ct.PlogRate(
    [(1013.25, ct.Arrhenius(1.2124e+16, -0.5779, 45491376.8)),
     (101325., ct.Arrhenius(4.9108e+31, -4.8507, 103649395.2)),
     (1013250., ct.Arrhenius(1.2866e+47, -9.0246, 166508556.0)),
     (10132500., ct.Arrhenius(5.9632e+56, -11.529, 220076726.4))])
```
This is fine in the case of `PlogRate`, where the base C++ reaction type is just `Reaction` but presents a problem for the reaction types where we still need C++ specializations of the `Reaction` class, as there's no information about the rate type available in the initial constructor.

What I'm wondering is whether we need this second constructor, where no rate information is provided. Would it make sense require specifying at least an empty `ReactionRate` object of the correct type here, e.g.
```python
R = ct.Reaction({"H2": 1, "O2": 1}, {"OH": 2}, rate=ct.PlogRate())
```
I'm also wondering if there's a good way to clean up the split between specifying the reactants and products separately versus specifying the reaction equation in these constructors.

**Checklist**

- [x] The pull request includes a clear description of this code change
- [x] Commit messages have short titles and reference relevant issues
- [x] Build passes (`scons build` & `scons test`) and unit tests address code coverage
- [x] Style & formatting of contributed code follows [contributing guidelines](https://github.com/Cantera/cantera/blob/main/CONTRIBUTING.md)
- [x] The pull request is ready for review

### CODE REVIEW

**Comment by ischoegl** [comment]
@speth ... I am overall in favor of this change, which moves along discussions we had during earlier PR's. My main question was about timing: should this be adopted prior to 2.6, or completed for 3.0? My assumption was to remove the old framework first, which would make a lot of specializations obsolete (this applies more to Python than to C++). In C++ we can move forward, and I do believe that some of the specializations can be further removed.

Regarding
```Python
R = Reaction(
    equation="H2 + O2 <=> 2 OH",
    rate=PlogRate([(1013.25, Arrhenius(1.2124e+16, -0.5779, 45491376.8)),
          (101325., Arrhenius(4.9108e+31, -4.8507, 103649395.2)),
          (1013250., Arrhenius(1.2866e+47, -9.0246, 166508556.0)),
          (10132500., Arrhenius(5.9632e+56, -11.529, 220076726.4))]),
    kinetics=gas)
```
it was my impression that this should work already? (I believe this should be covered by `test_reaction.py`).

Regarding the second constructor
```Python
R = ct.Reaction({"H2": 1, "O2": 1}, {"OH": 2}, rate=ct.PlogRate())
```
we can certainly enforce this, but for some of the legacy reactions, there is no dedicated `rate` object. So at least to me, the two-parameter version mostly exists for historical reasons.

**Comment by ischoegl** [comment]
@speth … I ended up deleting my earlier comment as I’m curious what you’re coming up with. I believe that `Reaction` specializations can be removed completely in C++ relatively soon (and in Python after 2.6). From what I’m seeing over in #1181, this includes `InterfaceReactions` (Which means that `ReactionFactory` would be gone also). All we’d have to do is to track dimensionality, which is easy based on a `ReactionRate` when defined in the API (or phase in the context of YAML input).

**Comment by speth** [comment]
When I started this, my hope was that it would lead to enabling removal of all the `Reaction` specializations in  C++ and along with that the `ReactionFactory` class, but I'm not so sure about that right now - there's a fair bit of specialization related to handling of third bodies and efficiencies for `ThreeBodyReaction` and `FalloffReaction`, and I don't know if we want to actually move that up to the base `Reaction` class, and I don't think it would make sense to make the `ReactionRate` classes know anything about the reactants / products. I'd be interested in any ideas you have that might make this possible, though.

**Comment by ischoegl** [comment]
>  I don't think it would make sense to make the `ReactionRate` classes know anything about the reactants / products. 

Agreed. But for setup purposes it may be necessary to let them have a peek (without saving anything)

> I'd be interested in any ideas you have that might make this possible, though.

... sure! For interface reactions, I had to do a fair amount of parsing when it comes to `StickingRate` (I split up `SurfaceArrhenius` as keeping everything together made little sense to me). The hook I came up with is https://github.com/Cantera/cantera/blob/d211185afd6a310976cfe86d145bf9f5297eddfd/include/cantera/kinetics/ReactionRate.h#L114-L116
I am planning to use the same to eliminate `ElectrochemicalReaction` (and the double-parsing required for their detection).

FWIW, here's the (preliminary) implementation used for `StickingRate::setContext` ('lifted' from `InterfaceKinetics`)
https://github.com/Cantera/cantera/blob/d211185afd6a310976cfe86d145bf9f5297eddfd/src/kinetics/Coverage.cpp#L176-L245

This function is called when the reaction is attached to a `Kinetics` object.

**PS:** 

> there's a fair bit of specialization related to handling of third bodies and efficiencies for `ThreeBodyReaction` and `FalloffReaction`, and I don't know if we want to actually move that up to the base `Reaction` class.

The situation is not too dissimilar to coverage dependence for `InterfaceReaction`. For that case, I ended up handling these by `ReactionRate` specializations.

File: test/data/kineticsfromscratch.yaml [hunk 1]
```diff
@@ -24,6 +24,7 @@ phases:
 reactions:
 - equation: O + H2 <=> H + OH  # Reaction 1
   rate-constant: {A: 38.7, b: 2.7, Ea: 6260.0}
+  duplicate: true  # duplicate of Reaction 7
```
Comment (ischoegl) [inline test/data/kineticsfromscratch.yaml]: Do we need to re-think the approach used for checking of duplicate reactions? Are reaction expressions distinct if there is a mismatch of reaction rate type?

File: src/kinetics/Reaction.cpp [hunk 2]
```diff
@@ -200,6 +202,14 @@ void Reaction::setRate(shared_ptr<ReactionRate> rate)
...
         m_rate = rate;
     }
+
+    if (reactants.count("(+M)")) {
+        warn_deprecated("Chebyshev reaction equation", input, "Specifying '(+M)' "
+            "in the reaction equation for Chebyshev reactions is deprecated.");
+        // remove optional third body notation
+        reactants.erase("(+M)");
+        products.erase("(+M)");
+    }
```
Comment (ischoegl) [inline src/kinetics/Reaction.cpp:225]: Behavior should be specific to `ChebyshevRate3`; as an aside, I am wondering whether #1165 could be resolved by a similar 'fix' (both here and in the conversion scripts).

File: src/kinetics/ReactionFactory.cpp [hunk 3]
```diff
@@ -22,11 +23,20 @@ std::mutex ReactionFactory::reaction_mutex;
...
-        return new ElementaryReaction3(node, kin);
+    reg("reaction", [](const AnyMap& node, const Kinetics& kin) {
+        unique_ptr<Reaction> R(new Reaction());
+        R->setParameters(node, kin);
+        if (kin.nPhases()) {
+            R->setRate(newReactionRate(node, R->calculateRateCoeffUnits3(kin)));
+        } else {
+            R->setRate(newReactionRate(node));
+        }
+        return R.release();
```
Comment (ischoegl) [inline src/kinetics/ReactionFactory.cpp]: Could this behavior be absorbed by a 'smarter' constructor `Reaction(const AnyMap& node, const Kinetics& kin)`? I am cautiously hopeful that `ReactionFactory` will no longer be necessary once XML/CTI are removed.

File: interfaces/cython/cantera/examples/kinetics/blowers_masel.py [hunk 4]
```diff
@@ -23,16 +23,16 @@
 import matplotlib.pyplot as plt
 
 #Create an elementary reaction O+H2<=>H+OH
-r1 = ct.ElementaryReaction({'O':1, 'H2':1}, {'H':1, 'OH':1})
+r1 = ct.Reaction({"O":1, "H2":1}, {"H":1, "OH":1})
```
Comment (ischoegl) [inline interfaces/cython/cantera/examples/kinetics/blowers_masel.py]: I think that
```Python
r1 = ct.Reaction({"O":1, "H2":1}, {"H":1, "OH":1})
r1.rate = ct.ArrheniusRate(3.87e1, 2.7, 6260*1000*4.184)
```
could be combined to
```Python
r1 = ct.Reaction({"O":1, "H2":1}, {"H":1, "OH":1}, 
    ct.ArrheniusRate(3.87e1, 2.7, 6260*1000*4.184))
```
for clarity.

Same goes for instances below and similar occurrences in `test_kinetics.py`.

File: interfaces/cython/cantera/test/test_reaction.py [hunk 5]
```diff
@@ -707,7 +711,11 @@ def from_rate(self, rate):
     def from_parts(self):
         # create reaction rate object from parts
         orig = self.gas.reaction(self._index)
-        rxn = self._cls(orig.reactants, orig.products, legacy=self._legacy)
+        if self._rate_cls:
+            rxn = self._cls(orig.reactants, orig.products, rate=self._rate_cls(),
+                            legacy=self._legacy)
```
Comment (ischoegl) [inline interfaces/cython/cantera/test/test_reaction.py]: Should use `rate=self._rate_obj` directly.

File: interfaces/cython/cantera/test/test_reaction.py [hunk 6]
```diff
@@ -804,7 +815,7 @@ def test_no_rate(self):
         # check behavior for instantiation from keywords / no rate
         if self._rate_obj is None:
             return
-        rxn = self.from_rate(None)
+        rxn = self.from_rate(self._rate_cls() if self._rate_cls else None)
```
Comment (ischoegl) [inline interfaces/cython/cantera/test/test_reaction.py]: I assume this is necessary as an empty type is needed to check that uninitialized rate objects work correctly?

I actually think that this PR will make uninitialized `ReactionRate` objects unnecessary. I.e. if no rate is attached, evaluation should fail. Likewise, generation of partial YAML input (e.g. `equation` without rate information) can be removed as most of the YAML content is now attached to the rate object anyways.

File: interfaces/cython/cantera/reaction.pyx [hunk 7]
```diff
@@ -795,6 +795,44 @@ cdef class Reaction:
             if products:
                 self.products = products
 
+    def __init__(self, reactants=None, products=None, *, equation=None, rate=None,
+                 Kinetics kinetics=None, init=True, legacy=False, **kwargs):
```
Comment (ischoegl) [inline interfaces/cython/cantera/reaction.pyx]: Would it make sense to write this as
```Python
    def __init__(self, reactants=None, products=None, rate=None, *, equation=None, 
                 Kinetics kinetics=None, init=True, legacy=False, **kwargs):
```
so the default sequence matches that of the C++ call?

Also see comment in `blowers_masel.py`.

File: interfaces/cython/cantera/reaction.pyx [hunk 8]
```diff
@@ -795,6 +795,44 @@ cdef class Reaction:
...
+
+        if legacy:
+            return
+
+        if reactants and products and not equation:
+            equation = self.equation
```
Comment (ischoegl) [inline interfaces/cython/cantera/reaction.pyx]: This leaves the object in an interesting state. Should there be any tests? I.e. what does calling the `rate` property getter do?

File: interfaces/cython/cantera/reaction.pyx [hunk 9]
```diff
@@ -795,6 +795,44 @@ cdef class Reaction:
...
+
+        if init and equation and kinetics:
+            rxn_type = self._reaction_type
+            spec = {"equation": equation, "type": rxn_type}
+            if isinstance(rate, dict):
+                # Arrhenius-like rates
+                spec["rate-constant"] = rate
+                rate = None
```
Comment (ischoegl) [inline interfaces/cython/cantera/reaction.pyx]: this should use `ReactionRate.from_dict` instead.

File: interfaces/cython/cantera/reaction.pyx [hunk 10]
```diff
@@ -795,6 +795,44 @@ cdef class Reaction:
...
+            if isinstance(rate, ReactionRate):
+                self.rate = rate
+                rate = None
+            elif rate is not None:
+                try:
+                    if isinstance(rate[0][1], Arrhenius):
+                        self.rate = PlogRate(rate)
+                        rate = None
+                except (IndexError, TypeError, KeyError):
+                    pass
```
Comment (ischoegl) [inline interfaces/cython/cantera/reaction.pyx]: with the exception of a dictionary (see comment above) and the potential exception of a (YAML-compatible) string, I believe all other options should be removed. For `PlogRate`, it may make sense to convert the output from the nested list to a dictionary (an idea that I played with, but ultimately dropped for reasons of back-ward compatibility).

**Review by ischoegl — CHANGES_REQUESTED** [review]
@speth ... this looks very good as it does cut a substantial amount of clutter that accumulated as Cantera/enhancements#87 progressed: you were even able to untangle the Python API, which is something I though would have to be deferred!

I have a couple of relatively small comments. The main question I have is whether the instantiation of a `ReactionRate` object shouldn't be called from within `Reaction.h` rather than `ReactionFactory`, as I have some hopes that we can eliminate the latter eventually.

Regarding merges, I believe this can be safely merged around the same time as #1166, and I will build on this to rework #1181 (which would otherwise introduce a slew of new objects that would be immediately removed). 

Beyond, I think it may make sense to have another look at `ThirdBodyCalc` as this is the only obstacle I see to removing some of the remaining reaction specializations. Of course, some changes will also be necessary for the science section on Cantera/cantera-website :cry:

**PS:** ugh, it looks like all of my multi-line comments were converted to regular comments (I reviewed from VSCode this time).

File: test/data/kineticsfromscratch.yaml (hunk 1 above)
Comment (speth) [reply test/data/kineticsfromscratch.yaml]: Yes, but I think this is a bigger issue than this PR. I've created a new enhancement post detailing some of the thoughts that I've been entertaining on this topic (Cantera/enhancements#132).

File: src/kinetics/Reaction.cpp (hunk 2 above)
Comment (speth) [reply src/kinetics/Reaction.cpp:225]: Oops. I had that and lost it in a rebase option, I guess.

File: interfaces/cython/cantera/examples/kinetics/blowers_masel.py (hunk 4 above)
Comment (speth) [reply interfaces/cython/cantera/examples/kinetics/blowers_masel.py]: Are you suggesting that we make `rate` a required argument to the `Reaction` constructor? I certainly think that would simplify matters.

File: test/data/kineticsfromscratch.yaml (hunk 1 above)
Comment (ischoegl) [reply test/data/kineticsfromscratch.yaml]: I think the new enhancement request goes in the right direction, but for the time being, these reactions should not be labeled as a duplicate as the rate parameterizations differ (according to the 'traditional' approach)

File: interfaces/cython/cantera/examples/kinetics/blowers_masel.py (hunk 4 above)
Comment (ischoegl) [reply interfaces/cython/cantera/examples/kinetics/blowers_masel.py]: I wasn't necessarily suggesting this, but it would probably avoid undesirable edge cases that need to be caught otherwise.

File: interfaces/cython/cantera/test/test_reaction.py (hunk 6 above)
Comment (speth) [reply interfaces/cython/cantera/test/test_reaction.py]: Without this, the tests of the `__call__` method fail for the rate types where the `__call__` method takes two parameters, but the default `ReactionRate` object is an `ArrheniusRate` that only takes one.

Comment (ischoegl) [reply interfaces/cython/cantera/test/test_reaction.py]: Ah, ok. I think we can leave the handling of empty rate parameterizations for another time (but a comment should be added) … *edit: actually, requiring rate as a parameter should resolve this?*

File: interfaces/cython/cantera/reaction.pyx (hunk 8 above)
Comment (ischoegl) [reply interfaces/cython/cantera/reaction.pyx]: Per your comment above, requiring rate as a parameter would resolve this edge case

File: interfaces/cython/cantera/test/test_reaction.py (hunk 6 above)
Comment (speth) [reply interfaces/cython/cantera/test/test_reaction.py]: ~~Well, it would resolve it in the sense that `None` would not be a viable argument to `from_rate`.~~
I put in a comment in `from_rate` that I hope explains this well enough. I'm looking forward to getting rid of the "legacy" rates so we can reduce the complexity of both the actual implementations and this testing infrastructure.

File: interfaces/cython/cantera/reaction.pyx (hunk 8 above)
Comment (speth) [reply interfaces/cython/cantera/reaction.pyx]: Right now, this gives you a `nan`-initialized `ArrheniusRate` object. I think that ultimately stems from the fact that not specifying a reaction `type` implies an elementary/Arrhenius rate.

**Comment by ischoegl** [comment]
@speth ... I thought about this some more over the weekend, and ended up opening Cantera/enhancements#133.

Beyond, to unwind the remaining tasks pertaining to Cantera/enhancements#87, I believe this PR should go first, then #1184, then #1181, and finally whatever we decide for Cantera/enhancements#133. The outcome would be complete removal of `Reaction` specializations.

File: interfaces/cython/cantera/examples/kinetics/blowers_masel.py (hunk 4 above)
Comment (speth) [reply interfaces/cython/cantera/examples/kinetics/blowers_masel.py]: Well, I went ahead with implementing that option, so we'll see if that ends up improving things.

File: interfaces/cython/cantera/reaction.pyx (hunk 7 above)
Comment (speth) [reply interfaces/cython/cantera/reaction.pyx]: I've adopted this as the signature for all non-deprecated `Reaction` types, as part of making the `rate` argument required.

**Comment by speth** [comment]
@ischoegl - I think I've addressed all of your review comments on this.

**Review by ischoegl — APPROVED** [review]
@speth ... thank you for addressing the comments! This looks good to go - I believe #1184 should be next. As an aside, I am likewise looking forward to removing the legacy bits from the testing framework - but it does give me some peace of mind until the transition is complete.
